pkg/foxtrot/db_test.go

```diff
@@ -72,3 +73,141 @@ func TestCreateUserErr(t *testing.T) {
+	require.Error(t, err)
+}
+
+func TestCreateRoomErr(t *testing.T) {
+	db, err := newDB("")
+	require.NoError(t, err)
+	defer db.close()
+
+	err = db.createRoom(context.Background(), &Room{Name: ""})
+	require.Error(t, err) // room cannot be empty
+}
+
+func TestCreateQueryMessageSimple(t *testing.T) {
+	db, err := newDB("")
+	require.NoError(t, err)
+	defer db.close()
+
+	ctx := context.Background()
+	require.NoError(t, db.createRoom(ctx, &Room{Name: "kitchen"}))
+	require.NoError(t, db.createUser(ctx, &User{Name: "alice", passwordHash: "###"}))
+
+	m := Message{Content: "hi", Room: "kitchen", Author: "alice"}
+	err = db.createMessage(ctx, &m)
+	require.NoError(t, err)
+
+	messages, err := db.queryMessages(ctx, "kitchen", -1, -1)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(messages))
+
+	got := messages[0]
+	createdAt, err := time.Parse(time.RFC3339, got.CreatedAt)
+	require.NoError(t, err)
+	since := time.Since(createdAt)
+	require.True(t, since < time.Second)
+
+	want := &Message{ID: 1, Content: "hi", Room: "kitchen", Author: "alice", CreatedAt: got.CreatedAt}
+	require.Equal(t, want, got)
+
+	messages, err = db.queryMessages(ctx, "kitchen", 10, -1)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(messages))
+	require.Equal(t, want, messages[0])
+
+	messages, err = db.queryMessages(ctx, "kitchen", -1, 10)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(messages))
+	require.Equal(t, want, messages[0])
+
+	messages, err = db.queryMessages(ctx, "kitchen", 10, 10)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(messages))
+	require.Equal(t, want, messages[0])
+}
+
+func TestCreateQueryMessage(t *testing.T) {
+	db, err := newDB("")
+	require.NoError(t, err)
+	defer db.close()
+
+	ctx := context.Background()
+	require.NoError(t, db.createRoom(ctx, &Room{Name: "kitchen"}))
+	require.NoError(t, db.createRoom(ctx, &Room{Name: "shed"}))
+	require.NoError(t, db.createUser(ctx, &User{Name: "alice", passwordHash: "###"}))
+	require.NoError(t, db.createUser(ctx, &User{Name: "bob", passwordHash: "***"}))
+
+	messages := []Message{
+		{Content: "ouch", Room: "shed", Author: "bob"},
+		{Content: "hi", Room: "kitchen", Author: "bob"},
+		{Content: "hungry?", Room: "kitchen", Author: "alice"},
+		{Content: "🍪🧁", Room: "shed", Author: "alice"},
+	}
+	for _, message := range messages {
+		m := message
+		require.NoError(t, db.createMessage(ctx, &m))
+	}
+	got, err := db.queryMessages(ctx, "kitchen", -1, -1)
+	require.NoError(t, err)
+	require.Equal(t, 2, len(got))
+	require.Equal(t, "hungry?", got[0].Content)
+	require.Equal(t, "hi", got[1].Content)
+
+	got, err = db.queryMessages(ctx, "shed", 2, -1)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(got))
+	require.Equal(t, "ouch", got[0].Content)
+
+	got, err = db.queryMessages(ctx, "shed", -1, 1)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(got))
+	require.Equal(t, "🍪🧁", got[0].Content)
+}
+
+func TestCreateMessageErr(t *testing.T) {
+	db, err := newDB(":memory:?_foreign_keys=true")
```

Comment:
i think this should be the default dsn for all test cases. we should never test with foreign keys disabled.

Reply:
added it the foreign key constraint enforcement to schema setup as discussed and db creation to mustDB which takes dsn `":memory:"` (could also be just `""`)